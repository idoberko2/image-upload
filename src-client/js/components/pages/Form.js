// external
import React from 'react';
import { Formik } from 'formik';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

// utils
import sendFiles from '../../utils/sendFiles';
import validateFiles from '../../utils/validateFiles';
import getSeason from '../../utils/getSeason';

// components
import FirstStep from '../steps/FirstStep';
import SecondStep from '../steps/SecondStep';
import ThirdStep from '../steps/ThirdStep';
import StatusRow from '../steps/StatusRow';
import ErrorMark from '../icons/ErrorMark';

// styles
import { mq } from '../common/globalCss';
import smallIconCss from '../icons/smallIconCss';

const formInitialValues = {
    collection: '',
    files: [],
    galleryName: '',
    photographer: '',
    season: getSeason(),
};

let filesRef = null;

function handleFilesChange(setFieldValue, setFieldTouched, event) {
    const { files } = event.target;

    if (!files || files.length === 0) {
        return;
    }

    setFieldValue('files', files);
    setFieldTouched('files');
}

function validate({ files }) {
    const filesArray = Array.from(files);
    let errors = {};

    if (filesArray.length === 0) {
        errors.files = 'לא נבחרו קבצים';
    } else if (
        !window.FileReader ||
        !window.Blob ||
        !validateFiles(filesArray)
    ) {
        errors.files = 'לפחות אחד מהקבצים אינו תמונה';
    }

    return errors;
}

export const MyInnerForm = ({
    errors,
    handleChange,
    handleSubmit,
    isSubmitting,
    isValid,
    values,
    setFieldValue,
    setFieldTouched,
    status,
    touched,
}) => (
    <FormWrapper onSubmit={handleSubmit}>
        <FirstStep
            errors={errors}
            isDisabled={isSubmitting}
            handleChange={event => {
                handleChange(event);
                setFieldTouched(event.currentTarget.id, true);
            }}
            touched={touched}
            values={values}
        />
        <SecondStep
            files={values.files}
            filesError={errors.files}
            isFilesTouched={touched.files}
            isDisabled={isSubmitting}
            handleChange={event => {
                handleFilesChange(setFieldValue, setFieldTouched, event);
            }}
            inputRef={c => (filesRef = c)}
        />
        <ThirdStep
            isDisabled={!isValid || isSubmitting}
            isLoading={isSubmitting}
        />
        <SubmissionError error={status && status.error} />
    </FormWrapper>
);

const handleSubmit = (
    handleSuccessfulSubmission,
    values,
    { resetForm, setSubmitting, setStatus }
) =>
    sendFiles(values)
        .then(() => {
            handleSuccessfulSubmission(() => {
                filesRef.value = '';
                resetForm(formInitialValues);
                setSubmitting(false);
            });
        })
        .catch(err => {
            setStatus({ error: err });
            setSubmitting(false);
        });

const Form = ({ handleSuccessfulSubmission }) => (
    <Formik
        component={MyInnerForm}
        onSubmit={handleSubmit.bind(null, handleSuccessfulSubmission)}
        initialValues={formInitialValues}
        validate={validate}
    />
);

const FormWrapper = styled.form`
    margin: 0;
    padding: 0;
    width: 80%;
    max-width: 40em;

    display: grid;
    grid-template-columns: 100%;
    grid-template-rows: auto;
    grid-column-gap: 10%;
    grid-row-gap: 2em;

    ${mq} {
        grid-template-columns: 45% 45%;
    }
`;

export const SubmissionError = ({ error }) => (
    <StatusRow
        css={css`
            grid-column-start: 1;
            grid-row-start: 3;
            grid-column-end: 3;
            justify-content: center;
        `}
    >
        {!error ? null : (
            <>
                <ErrorMark css={smallIconCss} />
                <div>קרתה שגיאה במהלך העלאת הקבצים :(</div>
            </>
        )}
    </StatusRow>
);

export default Form;
