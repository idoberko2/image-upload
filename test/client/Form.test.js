// external
import React from 'react';
import { mount } from 'enzyme';

// components
import Form from '../../src-client/js/components/pages/Form';

const findInputById = (wrapper, id) =>
    wrapper.findWhere(n => n.name() === 'TextInput' && n.prop('id') === id);

describe('Form', () => {
    test('does not allow to submit in initial state', () => {
        const wrapper = mount(<Form />);
        const submitButton = wrapper.find(
            'button[data-testid="submit-button"]'
        );

        expect(submitButton.exists()).toBe(true);
        expect(submitButton.prop('disabled')).toBe(true);
    });

    test('handles single file selection correctly', () => {
        const wrapper = mount(<Form />);
        const uploaderInput = wrapper.find('input#uploader');

        expect(wrapper.find('[data-testid="uploader-status"]')).toHaveLength(0);

        uploaderInput.simulate('change', {
            target: {
                files: [
                    new File(['test1'], 'test1.jpg', { type: 'image/jpeg' }),
                ],
            },
        });

        expect(
            wrapper.find('[data-testid="uploader-status"]').text()
        ).toMatchSnapshot();
    });

    test('handles multipe file selection correctly', () => {
        const wrapper = mount(<Form />);
        const uploaderInput = wrapper.find('input#uploader');

        expect(wrapper.find('[data-testid="uploader-status"]')).toHaveLength(0);

        uploaderInput.simulate('change', {
            target: {
                files: Array(100)
                    .fill(0)
                    .map(
                        (_, i) =>
                            new File([`test${i}`], `test${i}.jpg`, {
                                type: 'image/jpeg',
                            })
                    ),
            },
        });

        expect(
            wrapper.find('[data-testid="uploader-status"]').text()
        ).toMatchSnapshot();
    });

    test('handles file selection cancellation correctly', () => {
        const wrapper = mount(<Form />);
        const uploaderInput = wrapper.find('input#uploader');

        expect(wrapper.find('[data-testid="uploader-status"]')).toHaveLength(0);

        uploaderInput.simulate('change', {
            target: {
                files: [
                    new File(['test1'], 'test1.jpg', { type: 'image/jpeg' }),
                ],
            },
        });

        // this is what happens when opens the file selection dialog and chooses "cancel"
        uploaderInput.simulate('change', { target: { files: [] } });

        expect(
            wrapper.find('[data-testid="uploader-status"]').text()
        ).toMatchSnapshot();
    });

    test('handles file types correctly', done => {
        const wrapper = mount(<Form />);
        const uploaderInput = wrapper.find('input#uploader');

        expect(wrapper.find('[data-testid="uploader-status"]')).toHaveLength(0);

        uploaderInput.simulate('change', {
            target: {
                files: [
                    new File(['test1'], 'test1.jpg', { type: 'image/jpeg' }),
                    new File(['test2'], 'test2.bmp', { type: 'image/bmp' }),
                    new File(['test3'], 'test3.gif', { type: 'image/gif' }),
                ],
            },
        });

        expect(
            wrapper.find('[data-testid="uploader-status"]').text()
        ).toMatchSnapshot();

        uploaderInput.simulate('change', {
            target: {
                files: [
                    new File(['test4'], 'test4.txt', { type: 'text/plain' }),
                ],
            },
        });

        setImmediate(() => {
            expect(
                wrapper.find('[data-testid="uploader-status"]').text()
            ).toMatchSnapshot();
            done();
        });
    });

    describe('enables and disables the submission button correctly', () => {
        const initializeValidForm = wrapper => {
            findInputById(wrapper, 'collection').simulate('change', {
                target: { name: 'collection', value: 'collection name' },
            });
            findInputById(wrapper, 'galleryName').simulate('change', {
                target: { name: 'galleryName', value: 'gallery name' },
            });
            findInputById(wrapper, 'season').simulate('change', {
                target: { name: 'season', value: '2018-2019' },
            });
            findInputById(wrapper, 'photographer').simulate('change', {
                target: { name: 'photographer', value: 'photographer name' },
            });

            wrapper.find('input#uploader').simulate('change', {
                target: {
                    name: 'files',
                    files: [
                        new File(['test1'], 'test1.jpg', {
                            type: 'image/jpeg',
                        }),
                    ],
                },
            });
        };

        test('initially submit is disabled', () => {
            const wrapper = mount(<Form />);

            expect(
                wrapper
                    .find('button[data-testid="submit-button"]')
                    .prop('disabled')
            ).toBe(true);
        });

        test('invalid collection disables submit button', done => {
            const wrapper = mount(<Form />);
            initializeValidForm(wrapper);

            expect(
                wrapper
                    .find('button[data-testid="submit-button"]')
                    .prop('disabled')
            ).toBe(false);

            findInputById(wrapper, 'collection').simulate('change', {
                target: { name: 'collection', value: '' },
            });

            setImmediate(() => {
                wrapper.update();

                expect(
                    wrapper
                        .find('button[data-testid="submit-button"]')
                        .prop('disabled')
                ).toBe(true);
                done();
            });
        });

        test('invalid gallery name disables submit button', done => {
            const wrapper = mount(<Form />);
            initializeValidForm(wrapper);

            expect(
                wrapper
                    .find('button[data-testid="submit-button"]')
                    .prop('disabled')
            ).toBe(false);

            findInputById(wrapper, 'galleryName').simulate('change', {
                target: { name: 'galleryName', value: '' },
            });

            setImmediate(() => {
                wrapper.update();

                expect(
                    wrapper
                        .find('button[data-testid="submit-button"]')
                        .prop('disabled')
                ).toBe(true);
                done();
            });
        });

        test('invalid season disables submit button', done => {
            const wrapper = mount(<Form />);
            initializeValidForm(wrapper);

            expect(
                wrapper
                    .find('button[data-testid="submit-button"]')
                    .prop('disabled')
            ).toBe(false);

            findInputById(wrapper, 'season').simulate('change', {
                target: { name: 'season', value: '' },
            });

            setImmediate(() => {
                wrapper.update();

                expect(
                    wrapper
                        .find('button[data-testid="submit-button"]')
                        .prop('disabled')
                ).toBe(true);
                done();
            });
        });

        test('invalid photographer disables submit button', done => {
            const wrapper = mount(<Form />);
            initializeValidForm(wrapper);

            expect(
                wrapper
                    .find('button[data-testid="submit-button"]')
                    .prop('disabled')
            ).toBe(false);

            findInputById(wrapper, 'photographer').simulate('change', {
                target: { name: 'photographer', value: '' },
            });

            setImmediate(() => {
                wrapper.update();

                expect(
                    wrapper
                        .find('button[data-testid="submit-button"]')
                        .prop('disabled')
                ).toBe(true);
                done();
            });
        });

        test('invalid file disables submit button', done => {
            const wrapper = mount(<Form />);
            initializeValidForm(wrapper);

            expect(
                wrapper
                    .find('button[data-testid="submit-button"]')
                    .prop('disabled')
            ).toBe(false);

            wrapper.find('input#uploader').simulate('change', {
                target: {
                    name: 'files',
                    files: [
                        new File(['test1'], 'test1.txt', {
                            type: 'text/plain',
                        }),
                    ],
                },
            });

            setImmediate(() => {
                wrapper.update();

                expect(
                    wrapper
                        .find('button[data-testid="submit-button"]')
                        .prop('disabled')
                ).toBe(true);
                done();
            });
        });
    });
});
