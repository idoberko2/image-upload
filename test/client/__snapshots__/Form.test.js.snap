// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form handles file selection cancellation correctly 1`] = `"קובץ אחד נבחר"`;

exports[`Form handles file types correctly 1`] = `"3 קבצים נבחרו"`;

exports[`Form handles file types correctly 2`] = `"לפחות אחד מהקבצים אינו תמונה"`;

exports[`Form handles multipe file selection correctly 1`] = `"100 קבצים נבחרו"`;

exports[`Form handles single file selection correctly 1`] = `"קובץ אחד נבחר"`;
